### ISSUE TITLE
need option to only have array vararg in ABI for use with JNA 

### ISSUE THREAD

**Issue by michelou** [lampepfl/dotty#9754]
## Minimized code

Simple example of [JNA](https://github.com/java-native-access/jna) interface mapping and usage. Minimized code includes 3 source files:

**[1]** `src\main\java\CLibraryJava.java` (*Java version*) :

```Java
package com.sun.jna.examples;

import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Platform;

public interface CLibraryJava extends Library {
    CLibraryJava INSTANCE = (CLibraryJava)
        Native.load((Platform.isWindows() ? "msvcrt" : "c"), CLibraryJava.class);

    void printf(String format, Object... args);
}
```

**[2]** `src\main\scala\CLibrary.scala` (*Scala version*) :

```Scala
package com.sun.jna.examples

import com.sun.jna.Library
import com.sun.jna.Native
import com.sun.jna.Platform

import scala.annotation.varargs

trait CLibrary extends Library {
  @varargs
  def printf(format: String, args: Any*): Unit
}

object CLibrary {
  val INSTANCE = {
    val libName = if (Platform.isWindows()) "msvcrt" else "c"
    Native.load(libName, classOf[CLibrary]).asInstanceOf[CLibrary]
  }
}
```

**[3]** `src\main\scala\HelloWorld.scala`:

```Scala
package com.sun.jna.examples

object HelloWorld {

  def main(args: Array[String]): Unit = {
    CLibraryJava.INSTANCE.printf("Hello, World\n")  // Java version => OK
    CLibrary.INSTANCE.printf("Hello, World\n")  // Scala version => Exception
  }

}
```

## Output

Using Scala 3 (Dotty nightly/release):

```scala
$ dotc -version
Dotty compiler version 0.28.0-bin-20200907-101e620-NIGHTLY-git-101e620 -- Copyright 2002-2020, LAMP/EPFL

$ dotr -classpath "%LOCALAPPDATA%\Temp\lib\jna-5.6.0.jar;%LOCALAPPDATA%\Temp\lib\jna-platform-5.6.0.jar;W:\jna-examples\hello\target\classes" com.sun.jna.examples.HelloWorld arg0 arg1 arg2 arg3 arg4
Hello, World
Exception in thread "main" java.lang.IllegalArgumentException: Unsupported argument type scala.collection.immutable.ArraySeq$ofRef at parameter 1 of function printf
        at com.sun.jna.Function.convertArgument(Function.java:629)
        at com.sun.jna.Function.invoke(Function.java:345)
        at com.sun.jna.Library$Handler.invoke(Library.java:265)
        at com.sun.proxy.$Proxy1.printf(Unknown Source)
        at com.sun.jna.examples.HelloWorld$.main(HelloWorld.scala:21)
        at com.sun.jna.examples.HelloWorld.main(HelloWorld.scala)
Error: Program execution failed (com.sun.jna.examples.HelloWorld)
```

[CFR](https://www.benf.org/other/cfr/) output for `CLibrary.scala`:

```java
// Compiled with scala3_0.28.0-bin-20200907-101e620-NIGHTLY-git-101e620 
package com.sun.jna.examples;

import com.sun.jna.Library;
import scala.collection.immutable.Seq;

public interface CLibrary
extends Library {
    public void printf(String var1, Seq<Object> var2);

    public static void printf$(CLibrary $this, String format, Object[] args) {
        $this.printf(format, args);
    }

    public void printf(String var1, Object ... var2);
}
```

[CFR](https://www.benf.org/other/cfr/) output for `HelloWorld.scala`:

```java
package com.sun.jna.examples;

import com.sun.jna.examples.CLibrary$;
import com.sun.jna.examples.CLibraryJava;
import java.io.Serializable;
import scala.runtime.ModuleSerializationProxy;
import scala.runtime.ScalaRunTime$;

public final class HelloWorld$
implements Serializable {
    public static final HelloWorld$ MODULE$ = new HelloWorld$();

    private HelloWorld$() {
    }

    private Object writeReplace() {
        return new ModuleSerializationProxy(HelloWorld$.class);
    }

    public void main(String[] args) {
        CLibraryJava.INSTANCE.printf("Hello, World\n", new Object[0]);
        CLibrary$.MODULE$.INSTANCE().printf("Hello, World\n", ScalaRunTime$.MODULE$.genericWrapArray(new Object[0]));
    }
}
```

Using Scala 2:

```scala
$ scalac -version
Scala compiler version 2.13.3 -- Copyright 2002-2020, LAMP/EPFL and Lightbend, Inc.

$ scala -classpath "%LOCALAPPDATA%\Temp\lib\jna-5.6.0.jar;%LOCALAPPDATA%\Temp\lib\jna-platform-5.6.0.jar;W:\jna-examples\hello\target\classes" com.sun.jna.examples.HelloWorld arg0 arg1 arg2 arg3 arg4
Hello, World
java.lang.IllegalArgumentException: Unsupported argument type scala.collection.immutable.Nil$ at parameter 1 of function printf
        at com.sun.jna.Function.convertArgument(Function.java:629)
        at com.sun.jna.Function.invoke(Function.java:345)
        at com.sun.jna.Library$Handler.invoke(Library.java:265)
        at com.sun.proxy.$Proxy1.printf(Unknown Source)
        at com.sun.jna.examples.HelloWorld$.main(HelloWorld.scala:21)
        at com.sun.jna.examples.HelloWorld.main(HelloWorld.scala)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at scala.reflect.internal.util.ScalaClassLoader.$anonfun$run$2(ScalaClassLoader.scala:105)
        at scala.reflect.internal.util.ScalaClassLoader.asContext(ScalaClassLoader.scala:40)
        at scala.reflect.internal.util.ScalaClassLoader.asContext$(ScalaClassLoader.scala:37)
        at scala.reflect.internal.util.ScalaClassLoader$URLClassLoader.asContext(ScalaClassLoader.scala:130)
        at scala.reflect.internal.util.ScalaClassLoader.run(ScalaClassLoader.scala:105)
        at scala.reflect.internal.util.ScalaClassLoader.run$(ScalaClassLoader.scala:97)
        at scala.reflect.internal.util.ScalaClassLoader$URLClassLoader.run(ScalaClassLoader.scala:130)
        at scala.tools.nsc.CommonRunner.run(ObjectRunner.scala:29)
        at scala.tools.nsc.CommonRunner.run$(ObjectRunner.scala:28)
        at scala.tools.nsc.ObjectRunner$.run(ObjectRunner.scala:43)
        at scala.tools.nsc.CommonRunner.runAndCatch(ObjectRunner.scala:35)
        at scala.tools.nsc.CommonRunner.runAndCatch$(ObjectRunner.scala:34)
        at scala.tools.nsc.MainGenericRunner.runTarget$1(MainGenericRunner.scala:70)
        at scala.tools.nsc.MainGenericRunner.run$1(MainGenericRunner.scala:91)
        at scala.tools.nsc.MainGenericRunner.process(MainGenericRunner.scala:103)
        at scala.tools.nsc.MainGenericRunner$.main(MainGenericRunner.scala:108)
        at scala.tools.nsc.MainGenericRunner.main(MainGenericRunner.scala)
Error: Program execution failed (com.sun.jna.examples.HelloWorld)
```

[CFR](https://www.benf.org/other/cfr/) output  for `CLibrary.scala`:

```java
// Compiled with scala2_2.13.3
package com.sun.jna.examples;

import com.sun.jna.Library;
import com.sun.jna.examples.CLibrary$;
import scala.collection.immutable.Seq;
import scala.reflect.ScalaSignature;
import scala.runtime.ScalaRunTime$;

@ScalaSignature(bytes="...")
public interface CLibrary
extends Library {
    public static CLibrary INSTANCE() {
        return CLibrary$.MODULE$.INSTANCE();
    }

    public static /* synthetic */ void printf$(CLibrary $this, String format, Object ... args) {
        $this.printf(format, args);
    }

    default public void printf(String format, Object ... args) {
        this.printf(format, ScalaRunTime$.MODULE$.genericWrapArray(args));
    }

    public void printf(String var1, Seq<Object> var2);
}
```

[CFR](https://www.benf.org/other/cfr/) output for `HelloWorld.scala`:

```java
// Compiled with scala2_2.13.3
package com.sun.jna.examples;

import com.sun.jna.examples.CLibrary$;
import com.sun.jna.examples.CLibraryJava;
import scala.collection.immutable.Nil$;

public final class HelloWorld$ {
    public static final HelloWorld$ MODULE$ = new HelloWorld$();

    public void main(String[] args) {
        CLibraryJava.INSTANCE.printf("Hello, World\n", new Object[0]);
        CLibrary$.MODULE$.INSTANCE().printf("Hello, World\n", Nil$.MODULE$);
    }

    private HelloWorld$() {
    }
}
```

## Expectation

No runtime exception (Scala 3 and Scala 2).

We further note two differences between Scala 3 and Scala 2 in the above [CFR](https://www.benf.org/other/cfr/) outputs :

1) the 2nd argument in offending `printf` call :
    - Scala 3 :  `ScalaRunTime$.MODULE$.genericWrapArray(new Object[0])`
    - Scala 2 : `Nil$.MODULE$`

2) the signature of the `printf` methods in interface `CLibrary` (*not related to the issue*) :
    - Scala 3 : `public void printf(String format, Object ... args)`
    - Scala 2 : `default public void printf(String format, Object ... args)`

**Comment by bishabosha** [comment]
minimised to
```scala
trait Foo { def varargs(t: AnyRef*): String }

class Handler extends java.lang.reflect.InvocationHandler {
  def invoke(obj: Object, method: java.lang.reflect.Method, args: Array[Object]): Object =
    val argStr = if (args == null) "" else " with args " + args.mkString("(", ",", ")")
    s"result of invoking $method$argStr" 
}

val foo = java.lang.reflect.Proxy.newProxyInstance(
  classOf[Handler].getClassLoader, Array(classOf[Foo]), new Handler).asInstanceOf[Foo]

@main def Test: Unit = 
  println(foo.varargs("hello")) // "result of invoking public abstract java.lang.String rs$line$2$Foo.varargs(scala.collection.immutable.Seq) with args (ArraySeq(hello))"
```

**Comment by bishabosha** [comment]
I dont think you can do much but continue to use the Java compiled version, unless we have a way to never emit `Seq` API for varargs

**Comment by smarter** [comment]
Yeah, I don't think there's anything sensible we can do here, Java varargs don't even exist in Java bytecode, they're just arrays, so I assume JNA is using reflection to handle it specially? I don't think the language should have a feature to cater to one library weird uses of reflection. On the other hand I see JNA supports custom type mapping according to its documentation so you may be able to map `Seq[Object]` (which is what `Any*` compiles to) to a C varargs?

**Comment by michelou** [comment]
@smarter @bishabosha Here is my response to the limited support of Dotty (Scala 3) for Java varargs:
1) The Dotty distribution does include the library `lib/scala-asm-7.3.1-scala-1.jar` ([ASM FAQ](https://asm.ow2.io/faq.html)); that allows me to surgically modify the Java bytecode generated by the Scala compiler.
2) I wrote a tiny program in Scala (~100 lines) to rewrite the two class files `CLibrary.class` (*definition* of method `printf`) and `HelloWorld$` (*invocation* of method `printf`).

[**`javap`**](https://docs.oracle.com/javase/7/docs/technotes/tools/windows/javap.html) output ***without the patch*** (3 method defintions):
```
Compiled from "CLibrary.scala"
public interface com.sun.jna.examples.CLibrary extends com.sun.jna.Library {
  public abstract void printf(java.lang.String, scala.collection.immutable.Seq<java.lang.Object>);
  public static void printf$(com.sun.jna.examples.CLibrary, java.lang.String, java.lang.Object[]);
  public abstract void printf(java.lang.String, java.lang.Object...);
}
```
```
Compiled from "HelloWorld.scala"
[...]
      14: getstatic     #44                 // Field com/sun/jna/examples/CLibrary$.MODULE$:Lcom/sun/jna/examples/CLibrary$;
      17: invokevirtual #47                 // Method com/sun/jna/examples/CLibrary$.INSTANCE:()Lcom/sun/jna/examples/CLibrary;
      20: ldc           #35                 // String Hello, World\n
      22: getstatic     #52                 // Field scala/runtime/ScalaRunTime$.MODULE$:Lscala/runtime/ScalaRunTime$;
      25: iconst_1
      26: anewarray     #4                  // class java/lang/Object
      29: dup
      30: iconst_0
      31: aconst_null
      32: aastore
      33: invokevirtual #56                 // Method scala/runtime/ScalaRunTime$.genericWrapArray:(Ljava/lang/Object;)Lscala/collection/immutable/ArraySeq;
      36: invokeinterface #61,  3           // InterfaceMethod com/sun/jna/examples/CLibrary.printf:(Ljava/lang/String;Lscala/collection/immutable/Seq;)V
      41: return
}
```
[**`javap`**](https://docs.oracle.com/javase/7/docs/technotes/tools/windows/javap.html) output ***with the patch*** (2 method definitions and same invocation as Java invocation):
```
Compiled from "CLibrary.scala"
public interface com.sun.jna.examples.CLibrary extends com.sun.jna.Library {
  public static void printf$(com.sun.jna.examples.CLibrary, java.lang.String, java.lang.Object[]);
  public abstract void printf(java.lang.String, java.lang.Object...);
}
```
```
Compiled from "HelloWorld.scala"
[...]
      14: getstatic     #44                 // Field com/sun/jna/examples/CLibrary$.MODULE$:Lcom/sun/jna/examples/CLibrary$;
      17: invokevirtual #47                 // Method com/sun/jna/examples/CLibrary$.INSTANCE:()Lcom/sun/jna/examples/CLibrary;
      20: ldc           #35                 // String Hello, World\n
      22: iconst_0
      23: anewarray     #4                  // class java/lang/Object
      26: invokeinterface #70,  3           // InterfaceMethod com/sun/jna/examples/CLibrary.printf:(Ljava/lang/String;[Ljava/lang/Object;)V
      31: return
}
```
Successful execution (*patched* class files):
```
Hello, World
Hello, World
```
In conclusion, providing an optional parameter to the [`varargs`](https://www.scala-lang.org/api/current/scala/annotation/varargs.html) annotation (with appropriate compiler handling) would solve the issue.
